sim/core/focus.go

```diff
@@ -2,41 +2,120 @@ package core
 
 import (
 	"fmt"
+	"math"
+	"slices"
 	"time"
 
 	"github.com/wowsims/cata/sim/core/proto"
 )
 
 // Time between focus ticks.
-const MaxFocus = 100.0
-const tickDuration = time.Second * 1
-const BaseFocusPerTick = 5.0
-
-// OnFocusGain is called any time focus is increased.
-type OnFocusGain func(sim *Simulation)
+const FocusTickDuration = time.Millisecond * 200
 
 type focusBar struct {
 	unit *Unit
 
-	focusPerTick float64
-
+	maxFocus     float64
 	currentFocus float64
+	baseFocusPerSecond float64
 
-	onFocusGain OnFocusGain
+	// List of focus levels that might affect APL decisions. E.g:
+	// [10, 15, 20, 30, 60, 85]
+	focusDecisionThresholds []int
+
+	// Slice with len == maxFocus+1 with each index corresponding to an amount of focus. Looks like this:
+	// [0, 0, 0, 0, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, ...]
+	// Increments by 1 at each value of focusDecisionThresholds.
+	cumulativeFocusDecisionThresholds []int
 
 	nextFocusTick time.Duration
+	isPlayer bool
+
+	// Multiplies focus regen from ticks.
+	FocusTickMultiplier float64
 
-	regenMetrics  *ResourceMetrics
-	refundMetrics *ResourceMetrics
+	regenMetrics        *ResourceMetrics
+	FocusRefundMetrics *ResourceMetrics
 }
 
-func (unit *Unit) EnableFocusBar(regenMultiplier float64, onFocusGain OnFocusGain) {
+func (unit *Unit) EnableFocusBar(maxFocus float64, baseFocusPerSecond float64, isPlayer bool) {
+	unit.SetCurrentPowerBar(FocusBar)
+
 	unit.focusBar = focusBar{
-		unit:          unit,
-		focusPerTick:  BaseFocusPerTick * regenMultiplier,
-		onFocusGain:   onFocusGain,
-		regenMetrics:  unit.NewEnergyMetrics(ActionID{OtherID: proto.OtherAction_OtherActionFocusRegen}),
-		refundMetrics: unit.NewEnergyMetrics(ActionID{OtherID: proto.OtherAction_OtherActionRefund}),
+		unit:                 unit,
+		maxFocus:            max(100, maxFocus),
+		FocusTickMultiplier: 1,
+		isPlayer: isPlayer,
+		baseFocusPerSecond: baseFocusPerSecond,
+		regenMetrics:         unit.NewFocusMetrics(ActionID{OtherID: proto.OtherAction_OtherActionFocusRegen}),
+		FocusRefundMetrics:  unit.NewFocusMetrics(ActionID{OtherID: proto.OtherAction_OtherActionRefund}),
```

Comment:
it looks like your auto formating is not running properly I see weird indents here

Reply:
Youre right, it wanted to use prettier for some reason because of a config file in root.. Fixing

Reply:
Fixed